rfcs/001 - EC2 schema.md

```diff
@@ -0,0 +1,113 @@
+
+The current instance module has ten columns. Adding the ones representing integration with other features will almost double its size.
+
+## Proposal
+
+The `instance` table is the core of the `aws_ec2` module and is the one that can grow more in complexity. The proposal is to keep adding the necessary columns to the `instance` table. It is better to have a unique table to look for instances than increase the complexity of the model, creating other tables and adding to the users the need to understand our model first. Eventually, we can add an `aws_ec2_simplified` module that can expose the basics to create an instance and leave the `aws_ec2` for advanced users.
+
+We can divide the feature list as follow:
+
+The direct implication in the `instance` table:
+
+- Launch templates
+- Spot instances
+- Dedicated hosts
+- Capacity reservations
+- Auto Scaling Groups
+- Elastic IP integration
+- Placement groups
+- Network interfaces
+
+No implication in the `instance` table:
+
+- Instance type - Availability zone relationship
+- AMIs management
+- EBS volume snapshots and life cycle
+
+The `Saving plans`, `Reserved instances` or `Scheduled instances` are not included in the list above since they seem to be budgeting solutions that need interaction with the user rather than infrastructure-related modules' features that could have any implication in the model.
+
+Let's start with the features without impact on the `instance` table. These features will be part of the `aws_ec2` module with their mappers. The Instance type table by availability zones will be a table with a composite primary key. It technically has a minimum impact in the `instance` table that will be a special FK to the composite primary key, but no UX model impact.
+
+The features with direct implications in the `instance` table can be divided into two groups: one group affects only the `aws_ec2` module, the other impacts more modules.
+
+Let's start with the first group, Placement groups and Network interfaces. These features will be part of their respective modules (probably `aws_vpc`?). Then, they will need to be integrated with the `instance` table in the `aws_ec2` module with an FK nullable column.
+
+The rest of the features will need to be implemented in the `aws_ec2` module and integrated with the `instance` table in the `aws_ec2` module with an FK nullable column.
+
+Spot instances and Auto Scaling Group can be updated at any point in time by the cloud. These will imply the development of the "safer sync faster apply" feature as a pre-requisite for the two.
+
+Launch templates, auto-scaling group requests and spot instances requests can be linked to an instance using the tags AWS assign with a specific pattern. Otherwise will be hard to link all the relationships due to the async nature of ASG and spot instances.
```

Comment:
I know for spot instance requests, instances come with a nullable spot request id field. Would that not be preferable to a tag? maybe it should live in `instance_metadata` since it is technically not mutable?

Reply:
Ok so, some  points here. 

1. I will rephrase better, but what I'm saying is that from the tags that AWS assign, is how we can know if an instance belongs to an ASG or a spot fleet or was launched based on a launch template. Based on this tag we can look and assign the respective FK.

2. The field you are saying for spot instance request id it is a tricky one and I think we should not trust it. When you ask for a spot fleet request, for example, it is created a spot  instance request that creates the instance, but that spot instance request (that is the one that comes in the instance field as spot request id) is then deleted from the spot requests table, because it belongs to the fleet, it was not a single request. What I want to say is that sometimes those ids can be ephemeral.

3. From this comment I've being thinking and you are completely right that some fields, specifically the FKs to spot requests and ASG should belong to `instance_metadata` and not to `instance` table.

I will take a look again and maybe draw some diagrams showing the relationships.